SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1

```diff
@@ -79,46 +116,66 @@ function Set-Secret
-        $sb.Append("Username: ").AppendLine($Secret.UserName)
-        $sb.Append("login: ").AppendLine($Secret.UserName)
+    
+    
+    if ($Secret -is [string]) {
+        $Secret = @{Notes = $Secret}
+    } elseif ($Secret -is [pscredential]) {
+        $Secret = @{Username = $Secret.Username; Password = $Secret.GetNetworkCredential().password}
     }
 
-    if($Secret.Password) {
-        $pass = $Secret.Password
-        if ($Secret -is [pscredential]) {
-            $pass = $Secret.GetNetworkCredential().password
-        } elseif ($pass -is [securestring]) {
-            $pass = $pass | ConvertFrom-SecureString
-        }
 
-        $sb.Append("Password: ").AppendLine($pass)
-        $sb.Append("password: ").AppendLine($pass)
-        $sb.Append("sudo_password: ").AppendLine($pass)
-    }
-
-    if($Secret.URL) {
-        $sb.Append("URL: ").AppendLine($Secret.URL)
-    }
+    
+    if ($Secret -is [hashtable]){
+        $SpecialKeys = @('Language', 'NoteType', 'Notes')
+        $Keys = $Secret.Keys.Where({$_ -notin $SpecialKeys })
 
-    if($Secret.Notes) {
-        $sb.AppendLine("Notes:").AppendLine($Secret.Notes)
-    }
+        foreach ($k in $Keys) {
+            if ($Secret.$k -is [securestring]) {
+                $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.$k)
+                $Secret.$k = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+            }
+            [Void]($sb.AppendLine("$($k): $($Secret.$k)"))
+        }
 
+        # Notes need to be on a new line
+        if ($null -ne $Secret.Notes) {
+            if ($Secret.Notes -is [securestring]) {
+                $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.Notes)
+                $Secret.Notes = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+            }
+            [Void]($sb.AppendLine("Notes: `n$($Secret.Notes)"))
+        }
+    } 
+    
     try {
-        $sb.ToString() | Invoke-lpass 'add', $Name, '--non-interactive'
-    } catch {
+        $res = Invoke-lpass 'show', '--sync=now', '--name', $Name -ErrorAction SilentlyContinue
+          $SecretExists = $null -ne $res 
+
+        if ($SecretExists) {
+            Write-Verbose "Editing secret" 
+            $sb.ToString() | Invoke-lpass 'edit', '--non-interactive', $Name
+        } else {
+            Write-Verbose "Adding new secret" 
+            $NoteTypeArgs = @()
+            if ($null -ne $Secret.NoteType) {
+                if ($Secret.NoteType -is [securestring]) { 
+                    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.NoteType)
+                    $Secret.NoteType = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+                }
+                 $NoteTypeArgs = @("--note-type=$($Secret.NoteType.ToLower().replace(' ','-'))") 
+            }
+            $sb.ToString() | Invoke-lpass 'add', $Name, '--non-interactive', $NoteTypeArgs
```

Comment:
I've attempted to use the Notes arguments here to resolve the issue of semi-column from the input... I was not successful but if it work, it would resolve the second moot point.

Reply:
Current brain think old brain needed a break... Not so complicated to address.